node/core/provisioner/src/disputes/tests.rs

```diff
@@ -0,0 +1,638 @@
+
+	// active which can't conclude onchain
+	let (first_idx, first_votes) =
+		input.add_active_disputes_which_cant_conclude_onchain(DISPUTES_PER_BATCH);
+
+	//concluded disputes unknown onchain
+	let (fifth_idx, fifth_votes) = input.add_concluded_disputes_unknown_onchain(DISPUTES_PER_BATCH);
+
+	// concluded disputes known onchain
+	let (fourth_idx, _) = input.add_concluded_disputes_known_onchain(DISPUTES_PER_BATCH);
+
+	// active disputes unknown onchain
+	let (second_idx, second_votes) = input.add_active_disputes_unknown_onchain(DISPUTES_PER_BATCH);
+
+	let metrics = metrics::Metrics::new_dummy();
+	let mut vote_queries: usize = 0;
+	test_harness(
+		|r| mock_overseer(r, &mut input, &mut vote_queries),
+		|mut tx: TestSubsystemSender| async move {
+			let lf = leaf();
+			let result = select_disputes(&mut tx, &metrics, &lf).await.unwrap();
+
+			assert!(!result.is_empty());
+
+			if cfg!(feature = "staging-client") {
+				assert_eq!(result.len(), 4 * DISPUTES_PER_BATCH);
+
+				// Naive checks that the result is partitioned correctly
+				let (first_batch, rest) : (Vec<DisputeStatementSet>, Vec<DisputeStatementSet>) = result.into_iter().partition(|d| d.session == first_idx);
+				assert_eq!(first_batch.len(), DISPUTES_PER_BATCH);
+
+				let (second_batch, rest) : (Vec<DisputeStatementSet>, Vec<DisputeStatementSet>) = rest.into_iter().partition(|d| d.session == second_idx);
+				assert_eq!(second_batch.len(), DISPUTES_PER_BATCH);
+
+				let (third_batch, rest) : (Vec<DisputeStatementSet>, Vec<DisputeStatementSet>)= rest.into_iter().partition(|d| d.session == third_idx);
+				assert_eq!(third_batch.len(), DISPUTES_PER_BATCH);
+
+				let (fifth_batch, rest) : (Vec<DisputeStatementSet>, Vec<DisputeStatementSet>) = rest.into_iter().partition(|d| d.session == fifth_idx);
+				assert_eq!(fifth_batch.len(), DISPUTES_PER_BATCH);
+
+				// Ensure there are no more disputes - fourth_batch should be dropped
+				assert_eq!(rest.len(), 0);
+
+				assert_eq!(
+					first_batch
+						.iter()
+						.map(|d| d.statements.len())
+						.fold(0, |acc, v| acc + v),
+					first_votes
+				);
+				assert_eq!(
+					second_batch
+						.iter()
+						.map(|d| d.statements.len())
+						.fold(0, |acc, v| acc + v),
+					second_votes
+				);
+				assert_eq!(
+					third_batch
+						.iter()
+						.map(|d| d.statements.len())
+						.fold(0, |acc, v| acc + v),
+					third_votes
+				);
+				assert_eq!(
+					fifth_batch
+						.iter()
+						.map(|d| d.statements.len())
+						.fold(0, |acc, v| acc + v),
+					fifth_votes
+				);
+			} else {
+				assert_eq!(result.len(), 5 * DISPUTES_PER_BATCH);
+				for i in 0..3 * DISPUTES_PER_BATCH {
+					assert_ne!(result.get(i).unwrap().session, fourth_idx);
+					assert_ne!(result.get(i).unwrap().session, fifth_idx);
+				}
+			}
+		},
+	);
+	assert!(vote_queries <= ACCEPTABLE_RUNTIME_VOTES_QUERIES_COUNT);
+}
+
+#[test]
+fn many_batches() {
+	const VALIDATOR_COUNT: usize = 100;
+	const DISPUTES_PER_PARTITION: usize = 1000;
```

Comment:
Generally I agree. I did it this way for the sole reason to check how votes fetching works with a lot of disputes - I want to be sure that we don't bombard the Runtime with requests.
I can do something similar with less test data but it will kill the idea of the test - to verify the actual behaviour with actual limits.

I've checked how much time the test takes on CI. Here is with `staging client` disabled:
```
 running 7 tests
test tests::select_candidates::can_succeed ... ok
test tests::select_availability_bitfields::each_corresponds_to_an_occupied_core ... ok
test tests::select_availability_bitfields::more_set_bits_win_conflicts ... ok
test tests::select_availability_bitfields::not_more_than_one_per_validator ... ok
test tests::select_availability_bitfields::more_complex_bitfields ... ok
test tests::select_candidates::selects_correct_candidates ... ok
test tests::select_candidates::selects_max_one_code_upgrade ... ok

test result: ok. 7 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out;      finished in 0.12s 
```
And with `staging client` enabled (including the two tests with large test data):
```
running 14 tests
test disputes::with_staging_api::tests::partitioning_doubled_onchain_vote ... ok
test disputes::with_staging_api::tests::partitioning_duplicated_dispute ... ok
test disputes::with_staging_api::tests::should_keep_vote_behaves ... ok
test disputes::with_staging_api::tests::partitioning_happy_case ... ok
test tests::select_candidates::can_succeed ... ok
test tests::select_availability_bitfields::each_corresponds_to_an_occupied_core ... ok
test tests::select_availability_bitfields::more_set_bits_win_conflicts ... ok
test tests::select_availability_bitfields::not_more_than_one_per_validator ... ok
test disputes::with_staging_api::tests::normal_flow ... ok
test tests::select_availability_bitfields::more_complex_bitfields ... ok
test tests::select_candidates::selects_correct_candidates ... ok
test tests::select_candidates::selects_max_one_code_upgrade ... ok
test disputes::with_staging_api::tests::many_batches ... ok
test disputes::with_staging_api::tests::votes_above_limit ... ok

test result: ok. 14 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.37s
```
I think the delay is negligible - 7 tests for 0.12s vs 14 tests for 0.37s.

Do you agree to keep the test or insist on refactoring it?

Reply:
This is tripling the time - I would not call that negligible. It is not necessarily a big deal, but I don't get why it makes a difference in tests if something happens let's say every 10th vote or every 1000th vote for example. To be clear, I am suggesting to do something like [here](https://github.com/paritytech/polkadot/blob/fd31407737999e8f1663799055e76b0efd3611cc/node/core/dispute-coordinator/src/spam_slots.rs#L41) - using different values for a constant in production code, depending on whether tests are enabled or not.

Reply:
I've reworked it as you requested.